File: agent-python/src/pi/agent/loop.py
# src/pi/agent/loop.py
from __future__ import annotations

import asyncio
from typing import Any, Callable

from ..ai.types import (
    AssistantMessage, Message, TextContent, ToolCall, ToolResultMessage, UserMessage,
    StreamDone, StreamError, TextDelta, ToolCallEnd, StreamStart,
)
from ..ai.base import LLMProvider, StreamOptions
from ..tools.base import ToolResult
from .context import AgentContext
from .events import (
    AgentEnd, AgentError, AgentStart,
    MessageEnd, MessageUpdate,
    ToolExecEnd, ToolExecStart, ToolExecUpdate,
    TurnEnd, TurnStart,
)

Emit = Callable[[Any], None]


async def agent_loop(
    user_message: str,
    ctx: AgentContext,
    provider: LLMProvider,
    emit: Emit,
    signal: asyncio.Event | None = None,
) -> list[Message]:
    """Execute one agent turn: user message → LLM → tool calls → repeat → done."""
    new_messages: list[Message] = []

    user_msg = UserMessage(content=user_message)
    ctx.messages.append(user_msg)
    new_messages.append(user_msg)

    emit(AgentStart())

    for _turn in range(ctx.config.max_turns):
        emit(TurnStart())

        opts = StreamOptions(
            api_key=ctx.config.api_key,
            model=ctx.config.model_id,
        )

        tool_calls: list[ToolCall] = []
        text_parts: list[str] = []
        assistant_msg: AssistantMessage | None = None

        try:
            async for event in provider.stream(ctx.messages, ctx.system_prompt, ctx.tools, opts):
                if signal and signal.is_set():
                    emit(AgentEnd(messages=new_messages))
                    return new_messages

                if isinstance(event, TextDelta):
                    text_parts.append(event.delta)
                    emit(MessageUpdate(delta=event.delta))
                elif isinstance(event, ToolCallEnd):
                    tool_calls.append(event.tool_call)
                elif isinstance(event, (StreamDone, StreamError)):
                    assistant_msg = event.message
        except asyncio.CancelledError:
            emit(AgentEnd(messages=new_messages))
            return new_messages

        if assistant_msg is None:
            content: list[Any] = []
            if text_parts:
                content.append(TextContent(text="".join(text_parts)))
            content.extend(tool_calls)
            assistant_msg = AssistantMessage(content=content)

        ctx.messages.append(assistant_msg)
        new_messages.append(assistant_msg)
        emit(MessageEnd(message=assistant_msg))

        if not tool_calls:
            emit(TurnEnd(message=assistant_msg, tool_results=[]))
            break

        # Execute tools
        tool_results: list[ToolResultMessage] = []
        tool_map = ctx.tool_map()

        if ctx.config.tool_execution == "parallel":
            tasks = [
                _execute_tool(tc, tool_map, emit, signal)
                for tc in tool_calls
            ]
            results = await asyncio.gather(*tasks)
            tool_results = [r for r in results if r is not None]
        else:
            for tc in tool_calls:
                r = await _execute_tool(tc, tool_map, emit, signal)
                if r is not None:
                    tool_results.append(r)

        for tr in tool_results:
            ctx.messages.append(tr)
            new_messages.append(tr)

        emit(TurnEnd(message=assistant_msg, tool_results=tool_results))
    else:
        emit(AgentError(message=f"Max turns ({ctx.config.max_turns}) exceeded"))

    emit(AgentEnd(messages=new_messages))
    return new_messages


async def _execute_tool(
    tc: ToolCall,
    tool_map: dict,
    emit: Emit,
    signal: asyncio.Event | None,
) -> ToolResultMessage | None:
    emit(ToolExecStart(id=tc.id, name=tc.name, args=tc.arguments))

    tool = tool_map.get(tc.name)
    if tool is None:
        result = ToolResult.err(f"Unknown tool: {tc.name}")
    else:
        def on_update(partial: str) -> None:
            emit(ToolExecUpdate(id=tc.id, partial=partial))

        try:
            result = await tool.execute(tc.id, tc.arguments, signal=signal, on_update=on_update)
        except Exception as e:
            result = ToolResult.err(f"Tool execution error: {e}")

    emit(ToolExecEnd(id=tc.id, name=tc.name, result=result, is_error=result.is_error))

    return ToolResultMessage(
        tool_call_id=tc.id,
        tool_name=tc.name,
        content=[TextContent(text=result.content)],
        is_error=result.is_error,
        details=result.details,
    )

File: agent-rust/src/agent/loop.rs
use std::sync::Arc;
use chrono::Utc;
use tokio::sync::mpsc;
use tokio_util::sync::CancellationToken;

use crate::agent::{executor, AgentContext, AgentEvent, ToolCall};
use crate::ai::{
    provider::Provider,
    types::{Message, MessageContent, Role, StreamEvent, StreamOptions},
};

/// Run the agent loop for one user message.
///
/// Emits events to `event_tx` for the TUI to display.
/// Runs until: LLM returns no tool calls, max_turns exceeded, or cancellation.
pub async fn run(
    user_msg: String,
    ctx: &mut AgentContext,
    provider: Arc<dyn Provider>,
    event_tx: mpsc::Sender<AgentEvent>,
    cancel: CancellationToken,
    max_turns: usize,
) -> anyhow::Result<()> {
    ctx.messages.push(Message {
        role: Role::User,
        content: vec![MessageContent::Text { text: user_msg }],
        timestamp: Utc::now().timestamp_millis(),
    });

    let _ = event_tx.send(AgentEvent::AgentStart).await;

    for _turn in 0..max_turns {
        let _ = event_tx.send(AgentEvent::TurnStart).await;

        let opts = StreamOptions {
            api_key: ctx.api_key.clone(),
            max_tokens: 8192,
        };

        let mut rx = provider.stream(
            &ctx.messages,
            &ctx.system_prompt,
            &ctx.tool_definitions(),
            opts,
        ).await?;

        let mut tool_calls: Vec<ToolCall> = Vec::new();
        let mut text_buf = String::new();

        // Consume the stream, collecting text and tool calls
        loop {
            tokio::select! {
                evt = rx.recv() => {
                    match evt {
                        Some(StreamEvent::TextDelta(delta)) => {
                            text_buf.push_str(&delta);
                            let _ = event_tx.send(AgentEvent::StreamChunk(delta)).await;
                        }
                        Some(StreamEvent::ToolCall { id, name, arguments }) => {
                            tool_calls.push(ToolCall { id, name, arguments });
                        }
                        Some(StreamEvent::Done { .. }) | None => break,
                        Some(StreamEvent::Error(e)) => {
                            let _ = event_tx.send(AgentEvent::Error(e.to_string())).await;
                            let _ = event_tx.send(AgentEvent::AgentEnd).await;
                            return Ok(());
                        }
                    }
                }
                _ = cancel.cancelled() => {
                    let _ = event_tx.send(AgentEvent::AgentEnd).await;
                    return Ok(());
                }
            }
        }

        // Build assistant message from what we received
        let mut assistant_content = Vec::new();
        if !text_buf.is_empty() {
            assistant_content.push(MessageContent::Text { text: text_buf });
        }
        for tc in &tool_calls {
            assistant_content.push(MessageContent::ToolCall {
                id: tc.id.clone(),
                name: tc.name.clone(),
                arguments: tc.arguments.clone(),
            });
        }

        if !assistant_content.is_empty() {
            ctx.messages.push(Message {
                role: Role::Assistant,
                content: assistant_content,
                timestamp: Utc::now().timestamp_millis(),
            });
        }

        // No tool calls → conversation turn is done
        if tool_calls.is_empty() {
            let _ = event_tx.send(AgentEvent::TurnEnd).await;
            break;
        }

        // Execute all tool calls (parallel)
        let results = executor::execute_parallel(
            tool_calls,
            &ctx.tools,
            &event_tx,
            cancel.clone(),
        ).await;

        // Append tool results to message history
        for r in results {
            ctx.messages.push(Message {
                role: Role::ToolResult,
                content: vec![MessageContent::ToolResult {
                    tool_call_id: r.call_id,
                    tool_name: r.tool_name,
                    content: r.content,
                    is_error: r.is_error,
                }],
                timestamp: Utc::now().timestamp_millis(),
            });
        }

        let _ = event_tx.send(AgentEvent::TurnEnd).await;
        // Loop again — LLM sees tool results and continues
    }

    let _ = event_tx.send(AgentEvent::AgentEnd).await;
    Ok(())
}
